service/matching/version_rule_helpers.go

```diff
@@ -476,13 +478,34 @@ func dfs(curr string, visited, inStack map[string]bool, nodes map[string][]strin
 	return false
 }
 
-func FindAssignmentBuildId(rules []*persistencepb.AssignmentRule) string {
+func FindAssignmentBuildId(rules []*persistencepb.AssignmentRule, runId string) (string, error) {
+	rampThreshold := -1.
+	var err error
 	for _, r := range rules {
 		if r.GetDeleteTimestamp() != nil {
 			continue
 		}
-		// TODO: implement filter and ramp
-		return r.GetRule().GetTargetBuildId()
+		if ramp := r.GetRule().GetPercentageRamp(); ramp != nil {
+			if rampThreshold == -1. {
+				rampThreshold, err = calcRampThreshold(runId)
+				if err != nil {
+					return "", err
+				}
+			}
+			if float64(ramp.GetRampPercentage()) <= rampThreshold {
+				continue
+			}
+		}
+		return r.GetRule().GetTargetBuildId(), nil
+	}
+	return "", nil
+}
+
+func calcRampThreshold(id string) (float64, error) {
+	h := fnv.New32a()
+	_, err := h.Write([]byte(id))
+	if err != nil {
+		return -1, err
 	}
-	return ""
+	return float64(h.Sum32()) / math.MaxUint32, nil
```

Comment:
Looking at this function and the `TestCalcRampThresholdUniform`, I'm not understanding why to divide by `math.MaxUint32`? Would appreciate it if you could explain. My understanding is that `calcRampThreshold` should take in a run id, and return a float between [0, 100), is that correct?